File: plugins/powerbi-to-sigma/skills/powerbi-to-sigma/scripts/lib/pbi_viz_kind.py
#!/usr/bin/env python3
"""pbi_viz_kind.py — Python mirror of lib/pbi_viz_kind.rb for the extractors.

Resolves a raw Power BI ``visualType`` to a Sigma element kind + ROLE CLASS from
the SAME two catalogs the Ruby builder reads, so the two maps cannot drift:

    refs/catalogs/viz-kind.json       native PBI visualTypes (exact match)
    refs/catalogs/custom-visual.json  third-party/AppSource visuals (regex match)

This replaces the extractors' old ``VISUAL_KIND.get(vtype, "bar")`` dict, whose
silent default coerced any unrecognized visual into a bar chart (measured on 4
real customer .pbix files: 21 third-party date-picker SLICERS became bar charts).

All prose — per-row guidance plus the unknown/heuristic templates — lives in the
catalog JSON, so this file holds only resolution logic. Stdlib only.
"""
import json
import os
import re

FUNCTIONAL_ROLES = ("control", "kpi", "chart", "table")
NO_TARGET_ROLES = ("decoration", "unsupported")


def functional(role_class):
    """True when losing this visual is a FUNCTIONAL regression (gate-worthy)."""
    return str(role_class or "") in FUNCTIONAL_ROLES


def default_catalog_dir(script_file):
    """scripts/<any>.py or scripts/lib/<any>.py -> <skill>/refs/catalogs"""
    here = os.path.dirname(os.path.abspath(script_file))
    for up in ("..", os.path.join("..", "..")):
        cand = os.path.normpath(os.path.join(here, up, "refs", "catalogs"))
        if os.path.isdir(cand):
            return cand
    return os.path.normpath(os.path.join(here, "..", "refs", "catalogs"))


class Catalog:
    def __init__(self, catalog_dir):
        self.dir = catalog_dir
        with open(os.path.join(catalog_dir, "viz-kind.json")) as f:
            self.rows = json.load(f).get("rows") or []
        cvp = os.path.join(catalog_dir, "custom-visual.json")
        self.env = {}
        if os.path.exists(cvp):
            with open(cvp) as f:
                self.env = json.load(f)
        self.custom_rows = self.env.get("rows") or []
        self._by_type = {}
        for r in self.rows:
            for t in (r.get("pbi_visual_types") or []):
                self._by_type.setdefault(str(t).lower(), r)
        self._compiled = [(re.compile(cr["match"], re.I), cr)
                          for cr in self.custom_rows if cr.get("match")]
        self._slicer = re.compile(
            self.env.get("slicer_hint_pattern") or r"slicer|filter|picker", re.I)
        self._date = re.compile(
            self.env.get("date_hint_pattern") or r"date|timeline|calendar", re.I)

    def resolve(self, visual_type):
        """-> normalized row dict, or None when nothing matches."""
        t = str(visual_type or "").strip()
        if not t:
            return None
        r = self._by_type.get(t.lower())
        if r:
            return self._normalize(r, t, "viz-kind")
        for rx, cr in self._compiled:
            if rx.search(t):
                return self._normalize(cr, t, "custom-visual")
        return self._heuristic(t)

    def resolve_or_guidance(self, visual_type):
        """Never None — every caller gets a role_class + guidance to record."""
        return self.resolve(visual_type) or self.unknown_row(visual_type)

    def unknown_row(self, token):
        tmpl = self.env.get("unknown_guidance") or (
            "Unrecognized visual '{token}' — no Sigma kind was assumed; recorded as "
            "unsupported rather than silently rendered as a chart.")
        return {"source": str(token), "role_class": "unsupported", "sigma": None,
                "sigma_target": None, "builder_kind": None, "catalog": "none",
                "approximate": False, "visual_type": str(token),
                "guidance": tmpl.replace("{token}", str(token))}

    def _heuristic(self, t):
        """An unlisted visual whose NAME says it filters -> a control. Conservative
        by design: a false control is recoverable, a false bar chart silently
        deletes a page's filter."""
        if not self._slicer.search(t):
            return None
        kind = "date-range" if self._date.search(t) else "list"
        tmpl = self.env.get("heuristic_guidance") or (
            "Unlisted visual '{token}' looks like a filter — emitted as a Sigma {kind} control.")
        return {"source": t, "role_class": "control", "sigma": kind, "sigma_target": kind,
                "builder_kind": "control", "catalog": "heuristic", "approximate": False,
                "visual_type": t,
                "guidance": tmpl.replace("{token}", t).replace("{kind}", kind)}

    def _normalize(self, r, token, which):
        """The two catalogs carry the two names in DIFFERENT fields:
        viz-kind.json      `source` = the coarse token the Ruby builder switches on
                           (SIGMA_KIND[source] = sigma); `sigma` = the Sigma kind.
        custom-visual.json `sigma`  = that coarse token; `sigma_target` = Sigma kind.
        Getting this backwards emits sigma_kind='bar-chart' where the builder wants
        'bar' — caught by test-extract-viz-signals.py."""
        role = str(r.get("role_class") or "")
        builder = r.get("source") if which == "viz-kind" else r.get("sigma")
        sigma = r.get("sigma_target") or r.get("sigma")
        if role == "control" and (sigma is None or sigma == "control"):
            sigma = "list"
        if role in NO_TARGET_ROLES:
            sigma = None
            builder = None      # inert: the builder routes on role_class
        approx = token.lower() in [str(x).lower()
                                   for x in (r.get("approximate_types") or [])]
        return {"source": r.get("source"), "role_class": role, "sigma": sigma,
                "sigma_target": r.get("sigma_target"), "builder_kind": builder,
                "guidance": r.get("guidance"), "notes": r.get("notes"),
                "approximate": approx, "catalog": which, "visual_type": token}


def load(catalog_dir):
    return Catalog(catalog_dir)

File: plugins/powerbi-to-sigma/skills/powerbi-to-sigma/scripts/lib/pbi_viz_kind.rb
# frozen_string_literal: true
#
# PbiVizKind — resolve a raw Power BI `visualType` to a Sigma element kind + ROLE
# CLASS, from the documentation-grounded catalogs (single source of truth):
#
#   refs/catalogs/viz-kind.json       native PBI visualTypes (exact match)
#   refs/catalogs/custom-visual.json  third-party/AppSource visuals (regex match)
#
# WHY THIS EXISTS (measured on 4 real customer .pbix files, 2026-07-30): the
# extractors mapped visualType -> kind with `VISUAL_KIND.get(vt, "bar")`, so ANY
# unrecognized type silently became a bar chart. That turned 21 third-party
# date-picker SLICERS into bar charts (every page lost its date filter), plus 4
# modern `cardVisual` cards and 5 decorative `shape`s. The builder warned, but
# recorded them as severity 'approximated' — which the coverage headline counts as
# CARRIED OVER, so no gate failed — with guidance that was wrong ("Sigma has no
# native Datepicker"; Sigma has a native date-range control).
#
# Two design rules follow, both enforced by test-viz-kind-catalog.rb:
#   1. NO SILENT DEFAULT. An unresolved token returns nil; `resolve_or_guidance`
#      yields role_class 'unsupported' + actionable guidance and NO sigma target.
#   2. ROLE CLASS decides severity. Losing a `control` (the page lost its filter),
#      `kpi`, `chart` or `table` is a FUNCTIONAL loss the gate can fail on; losing
#      a `decoration` or `text` is cosmetic. `functional?` is that test. `image`
#      is ALSO non-functional (a missing logo is cosmetic, not a data loss) but
#      is its OWN role_class, distinct from `decoration`: an image visual DOES
#      build a real Sigma element when --image-map supplies a hosted URL, and
#      has its own dedicated coverage entry when it can't — unlike a shape,
#      which never builds anything. Conflating the two made every image visual
#      return nil before ever reaching its build logic (review-caught regression).
#
# ALL prose (per-row guidance, the unknown/heuristic templates) lives in the
# catalog JSON, never in code — so this resolver and its Python mirror
# (lib/pbi_viz_kind.py, used by the extractors) stay thin and cannot drift in
# wording. Resolution order: viz-kind exact -> custom-visual regex -> generic
# slicer/filter heuristic -> nil. Pure + stdlib-only.
require 'json'

module PbiVizKind
  module_function

  FUNCTIONAL_ROLES = %w[control kpi chart table].freeze
  NO_TARGET_ROLES  = %w[decoration unsupported].freeze

  # True when losing this visual is a FUNCTIONAL regression (gate-worthy).
  def functional?(role_class)
    FUNCTIONAL_ROLES.include?(role_class.to_s)
  end

  def load(catalog_dir)
    Catalog.new(catalog_dir)
  end

  # Fallback templates if custom-visual.json is absent (viz-kind-only install).
  DEFAULT_UNKNOWN_GUIDANCE =
    "Unrecognized visual '{token}' — no Sigma kind was assumed; recorded as " \
    'unsupported rather than silently rendered as a chart. Inspect what it binds and ' \
    'what it does, then add a row to refs/catalogs/custom-visual.json.'

  class Catalog
    attr_reader :rows, :custom_rows

    def initialize(dir)
      @dir = dir
      @rows = JSON.parse(File.read(File.join(dir, 'viz-kind.json')))['rows'] || []
      cvp = File.join(dir, 'custom-visual.json')
      @env = File.exist?(cvp) ? JSON.parse(File.read(cvp)) : {}
      @custom_rows = @env['rows'] || []
      @by_type = {}
      @rows.each do |r|
        (r['pbi_visual_types'] || []).each { |t| @by_type[t.to_s.downcase] ||= r }
      end
      @compiled = @custom_rows.select { |cr| cr['match'] }
                              .map { |cr| [Regexp.new(cr['match'], Regexp::IGNORECASE), cr] }
      @slicer_hint = Regexp.new(@env['slicer_hint_pattern'] || 'slicer|filter|picker', Regexp::IGNORECASE)
      @date_hint   = Regexp.new(@env['date_hint_pattern'] || 'date|timeline|calendar', Regexp::IGNORECASE)
    end

    # Raw PBI visualType -> normalized row hash, or nil when nothing matches.
    # Keys: source, role_class, sigma (concrete Sigma kind), sigma_target,
    #       builder_kind (coarse token the builder switches on), guidance,
    #       approximate (data-preserving substitution), catalog, visual_type.
    def resolve(visual_type)
      t = visual_type.to_s.strip
      return nil if t.empty?
      if (r = @by_type[t.downcase])
        return normalize(r, t, 'viz-kind')
      end
      if (pair = @compiled.find { |rx, _| rx.match?(t) })
        return normalize(pair[1], t, 'custom-visual')
      end
      heuristic(t)
    end

    # Never nil — every caller gets a role_class + guidance to record.
    def resolve_or_guidance(visual_type)
      resolve(visual_type) || unknown_row(visual_type)
    end

    def unknown_row(token)
      tmpl = @env['unknown_guidance'] || PbiVizKind::DEFAULT_UNKNOWN_GUIDANCE
      { 'source' => token.to_s, 'role_class' => 'unsupported', 'sigma' => nil,
        'sigma_target' => nil, 'builder_kind' => nil, 'catalog' => 'none',
        'approximate' => false, 'visual_type' => token.to_s,
        'guidance' => tmpl.gsub('{token}', token.to_s) }
    end

    private

    # An unlisted visual whose NAME says it filters -> a control. Deliberately
    # conservative: a false control is recoverable; a false bar chart silently
    # deletes a page's filter.
    def heuristic(t)
      return nil unless @slicer_hint.match?(t)
      kind = @date_hint.match?(t) ? 'date-range' : 'list'
      tmpl = @env['heuristic_guidance'] ||
             "Unlisted visual '{token}' looks like a filter — emitted as a Sigma {kind} control."
      { 'source' => t, 'role_class' => 'control', 'sigma' => kind, 'sigma_target' => kind,
        'builder_kind' => 'control', 'catalog' => 'heuristic', 'approximate' => false,
        'visual_type' => t,
        'guidance' => tmpl.gsub('{token}', t).gsub('{kind}', kind) }
    end

    def normalize(r, token, which)
      role = r['role_class'].to_s
      # The two catalogs carry the two names in DIFFERENT fields:
      #   viz-kind.json      `source` = the coarse token the Ruby builder switches
      #                      on (SIGMA_KIND[source] = sigma); `sigma` = Sigma kind.
      #   custom-visual.json `sigma`  = that coarse token; `sigma_target` = Sigma kind.
      # Getting this backwards emits sigma_kind:'bar-chart' where the builder
      # expects 'bar' — caught by test-extract-viz-signals.py.
      builder = which == 'viz-kind' ? r['source'] : r['sigma']
      # concrete Sigma kind: explicit sigma_target wins; a control with no target
      # defaults to a list control; decoration/unsupported get NO target so no
      # caller can accidentally render one.
      sigma = r['sigma_target'] || r['sigma']
      sigma = 'list' if role == 'control' && (sigma.nil? || sigma == 'control')
      if NO_TARGET_ROLES.include?(role)
        sigma = nil
        builder = nil                      # inert: the builder routes on role_class
      end
      approx = Array(r['approximate_types']).map { |x| x.to_s.downcase }.include?(token.downcase)
      { 'source' => r['source'], 'role_class' => role, 'sigma' => sigma,
        'sigma_target' => r['sigma_target'], 'builder_kind' => builder,
        'guidance' => r['guidance'], 'notes' => r['notes'], 'approximate' => approx,
        'catalog' => which, 'visual_type' => token }
    end
  end
end
